### ISSUE TITLE
Updated ios app, now it will not open, crashes on opening

### ISSUE THREAD

**Issue by njosephg** [mattrubin/authenticator#276]
App was working fine yesterday. 

I am on ios 10.3.2. After the update day yesterday, I can not use the app and not use the authenticator for my various logins, need a fix ASAP!

**Comment by vjl** [comment]
I'm on the same iOS version and also having the same crash. Thankfully github has a backup 2FA so I could login to leave this comment! But I am locked out of several that don't offer backup codes or SMS backups.

**Comment by mattrubin** [comment]
I'm sorry this crash is preventing you from using the app! I'll do my best to get this resolved as soon as possible. I've been trying to diagnose your issue, but unfortunately, I haven't been able to reproduce this crash in any of my testing. There's a couple of things to check that might help:

 - You mentioned that you're on iOS 10.3.2. What version of the iPhone (or other iOS device) are you using?
 - It's difficult to determine the cause of a crash without a crash report, and I haven't received any crash reports via Apple. It's possible that your phone might not be configured to send reports to the developers when an app crashes. To enable this, open the Settings app and select "Privacy", then scroll down and select "Analytics". If you turn on the switches next to "Share iPhone Analytics" and then "Share with App Developers", then the next time the app crashes, your phone will send information about the cause of that crash to Apple, and Apple will send that information to the app's developers. Can you please turn those switches on, or let me know if they were already turned on?
 - Have you tried turning your phone off and on again? It's a dumb solution, but sometimes it can help solve issues like this one.

**Comment by njosephg** [comment]
Yes i'm currently on a iPhone SE ios v.10.3.2

I just turned on my analytics and ran a few crashes for you.

Yeah rebooted a few times.

**Comment by vjl** [comment]
Apologies for the lag. I'm actually on iOS 10.3.3 not 10.3.2 as the OP is. Running a 128 GB 6s

I've had crash reports on for a long time. Generated 10 crashes today so far.  I can send you any/all of them privately if you'd like. 

I believe I restarted but I'll do that now to be doubly sure. The app has worked so well for so long and lives on my home screen. Hopefully we'll be able to provide enough data for you to reproduce or fix it. :)

**Comment by vjl** [comment]
I can confirm that a reboot does not solve the issue. And I've had my privacy settings set to send crash reports to both Apple and 3rd party devs since installing iOS 10. I guess Apple lags in forwarding them to you. :(

**Comment by njosephg** [comment]
I just fixed my issue by updating to the newest ios version 11.3.1. App is working again.

**Comment by vjl** [comment]
I won't be able to upgrade to iOS 11 for quite some time, unfortunately.

**Comment by paulscathedral** [comment]
Yeah, got the same problem. Hope you will fix it. I just don't wonna lose all my auth tokens. (

**Comment by paulscathedral** [comment]
5s, 10.3.3
The application minimizes all time. This is not a classic crash.

**Comment by mattrubin** [comment]
Thanks for all the additional info! I've paused the rollout of the new version, so hopefully this issue won't hit too many more people before we can get a fix out.


@njosephg: The fact that upgrading to iOS 11 fixed the issue on your phone helps me narrow down possible causes. Thanks for trying that!

@vjl: There seems to be quite a lag in Apple forwarding crash reports. If you could send a few of the crash reports manually to authenticator@mattrubin.me, I'd appreciate it!

@paulscathedral: your tokens are saved securely in the iOS keychain. The app crashing shouldn't have any affect on the storage of the tokens. I'll get this fixed as soon as possible so you can see your tokens again!

**Comment by vjl** [comment]
Crash reports are on their way!

**Comment by mattrubin** [comment]
Thanks for the crash reports, @vjl! The cause of the crash is an error encountered when retrieving token data from the iOS keychain when the app launches. It seems that in addition to the valid tokens, the keychain is also returning some additional bad data which the app doesn't know how to handle. I'm not sure why iOS 10.3 is having keychain issue when iOS 11 is not, but [a recent change](https://github.com/mattrubin/OneTimePassword/pull/161) to how the app handles keychain errors has likely turned this iOS 10.3 bug from something that was ignored to something that caused a crash.

I'm going to release a new beta version of the app which handles these errors in a way that will avoid crashing and will provide additional information about the underlying cause of the issue. If any of you are willing to test this new version, please send an email to authenticator@mattrubin.me, and I'll send you back an invitation to download the beta build.

**Comment by 1r2** [comment]
Same crash, iOS 10.3.3, iPhone 9,2.

I have root access of my phone, is there a way to copy those auth token out before the fix come out?

Thank you.

**Comment by bookwho** [comment]
iPhone 6, iOS 10.3, also having same crash problem. Hopefully this gets sorted soon. Can't login to several things now.

**Comment by mattrubin** [comment]
Beta version 2.0.4 is now available via TestFlight. If anyone else is encountering this crash and would like to help test the fix, please send an email to authenticator@mattrubin.me.

**Comment by mattrubin** [comment]
Version 2.0.4 of the app, which includes the fix for this crash, has been sent to Apple for review. I have requested an expedited review process, and configured the app to release the new version on the App Store immediately upon Apple's approval.

My apologies to all of you for the mistakes that led to this crash bug. I appreciate your patience while we worked to fix it, and am very grateful to those who helped to track down the issue and test the fix.
